sync/src/types/mod.rs

```diff
@@ -1603,17 +1583,69 @@ impl SyncState {
     // where T: Iterator<Item=Byte32>,
     // for<'a> &'a T: Iterator<Item=&'a Byte32>,
     pub fn mark_as_known_txs(&self, hashes: impl Iterator<Item = Byte32> + std::clone::Clone) {
+        let mut unknown_tx_hashes = self.unknown_tx_hashes.lock();
+        let mut tx_filter = self.tx_filter.lock();
+
+        for hash in hashes {
+            unknown_tx_hashes.remove(&hash);
+            tx_filter.insert(hash);
+        }
+    }
+
+    pub fn pop_ask_for_txs(&self) -> HashMap<PeerIndex, Vec<Byte32>> {
+        let mut unknown_tx_hashes = self.unknown_tx_hashes.lock();
+        let mut result: HashMap<PeerIndex, Vec<Byte32>> = HashMap::new();
+        let now = Instant::now();
+
+        if !unknown_tx_hashes
+            .peek()
+            .map(|(_tx_hash, priority)| priority.should_request(now))
+            .unwrap_or_default()
         {
-            let mut inflight_transactions = self.inflight_transactions.lock();
-            for hash in hashes.clone() {
-                inflight_transactions.pop(&hash);
+            return result;
+        }
+
+        while let Some((tx_hash, mut priority)) = unknown_tx_hashes.pop() {
+            if priority.should_request(now) {
+                if let Some(peer_index) = priority.next_request_peer() {
+                    result
+                        .entry(peer_index)
+                        .and_modify(|hashes| hashes.push(tx_hash.clone()))
+                        .or_insert_with(|| vec![tx_hash.clone()]);
+                    unknown_tx_hashes.push(tx_hash, priority);
+                }
+            } else {
+                unknown_tx_hashes.push(tx_hash, priority);
+                break;
             }
         }
+        result
+    }
 
-        let mut tx_filter = self.tx_filter.lock();
+    pub fn add_ask_for_txs(&self, peer_index: PeerIndex, tx_hashes: Vec<Byte32>) {
+        let mut unknown_tx_hashes = self.unknown_tx_hashes.lock();
+        if unknown_tx_hashes.len() >= MAX_UNKNOWN_TX_HASHES_SIZE {
+            return;
+        }
 
-        for hash in hashes {
-            tx_filter.insert(hash);
+        for tx_hash in tx_hashes
+            .into_iter()
+            .take(MAX_UNKNOWN_TX_HASHES_SIZE - unknown_tx_hashes.len())
+        {
+            match unknown_tx_hashes.entry(tx_hash) {
```

Comment:
This entry API looks weird, I think it might be better to use remove directly

Reply:
this fn is adding the tx hashes to the `unknown_tx_hashes`, not remove it from the `unknown_tx_hashes`

Reply:
> this fn `add_ask_for_txs` is adding the tx hashes to the `unknown_tx_hashes`, not remove it from the `unknown_tx_hashes`

I mean, since the entry must be cloned, it is better to remove it and add it and then plug it back in.

Reply:
the time complexity of remove and push it back will be O(log n) + O(n), clone and set_priority will be 0 since the order key of priority  (request_time) is not changed, I will try to benchmark the cost of clone vs time complexity according to your suggestion.

Reply:
I did a simple benchmark: https://gist.github.com/quake/d4a0f1bfa5b7b90f4e910039c87fb413, entry/clone is performed better than remove/push in all cases, you may try bench it with different tx_hashes size and peers size. 

This is a sample result on my local pc with tx_hashes size 150 and peers size 4:
```
entry_clone             time:   [50.812 us 51.001 us 51.199 us]        
remove_push             time:   [57.857 us 58.202 us 58.605 us]                        
```